smolyay/surrogate.py

```diff
@@ -1,546 +1,729 @@
-    and its derivative through :meth:`__call__(x)` and :meth:`derivative(x)`
-    at a given input. :meth:`gradient` takes the gradient of the
-    surrogate.
-    :meth:`_reset_surrogate` resets the surrogate's coefficients,
-    sampled points, original ``grids``(holds the grid points
-    and their corresponding basis functions depending on the
-    dimensionality and the generator), and ``data`` (function at
-    sampling points).
+    through :meth:`predict(x)`.
 
     Parameters
     ----------
-    domain: list
-        Domain of the function to be approximated.
-    grid_generator: IndexGridGenerator
-        Set of grids and their corresponding basis function.
-
-    Raises
-    ------
-    TypeError
-        Grids must be :class:`IndexGridGenerator`.
-        Domain must be dim x 2.
-
-    Example
-    -------
-    Consider a function :math:F(y) with one independent variable.
-    Basis function B_n(x) is chosen for the approximation and
-    three grid points and their corresponding basis functions
-    are generated in (-1, 1) domain:
-    (:math:x_0, :math:x_1, :math:x_2)
-    [:math:B_0(x), :math:B_1(x) :math:B_2(x)]
-    Based on the domain of the approximation, grid points of the basis
-    functions can be moved to the new domain,
-    resulting in transformed grid points:
-    (:math:y_0, :math:y_1, :math:y_2)
-    Then:
-    ..math::
-        \[
-        \begin{bmatrix}
-        B_0(x_0)& B_1(x_0) & B_2(x_0) \\
-        B_0(x_1)& B_1(x_1) & B_2(x_1) \\
-        B_0(x_2)& B_1(x_2) & B_2(x_2)
-        \end{bmatrix}
-        \begin{bmatrix}
-        C_0\\C_1\\C_2
-        \end{bmatrix}
-        =\begin{bmatrix}
-        F(y_0)\\F(y_1)\\F(y_2)
-        \end{bmatrix}
-        \]
-
-    Coefficients (x) can be obtained by solving linear equations,
-    since we have a full rank system of matrixes. The user can
-    select between 'lower upper decomposition' ('lu'), and
-    'inverse' ('inv') method for calculating the surrogate's (S)
-    coefficients. 'lstsq' solves the above equations through
-    least linear method.
-    S = C_0B_0(x) + C_1B_1(x) + C_2B_2(x)
+    domain: numpy.ndarray
+        the domain of the surrgoate
     """
 
-    def __init__(self, domain, grid_generator):
-        self.domain = domain
-        self.grid_generator = grid_generator
-        self._coefficients = None
-        self._grid = None
-        self._points = None
-        self._data = None
-
-    @property
-    def coefficients(self):
-        """list: coefficients of surrogate."""
-        if self._coefficients is not None:
-            return self._coefficients.tolist()
-        else:
-            return None
-
-    @property
-    def data(self):
-        """list: data at sampling grid points."""
-        if self._data is not None:
-            return self._data.tolist()
-        else:
-            return None
+    def __init__(self, domain):
+        self._domain = None
+        self._needs_fit = False
```

Comment:
This should be `True` (to indicate that fitting is needed). Make sure to update usage below so that it is set to `True` whenever a fit is needed, and to `False` after the fit is performed.
```suggestion
        self._needs_fit = True
```